From 120bf43284ea9d56a670d79ce28add6046b3db29 Mon Sep 17 00:00:00 2001
From: Alexander Gattin <alexander_gattin@xyratex.com>
Date: Sat, 23 Feb 2013 17:19:21 +0200
Subject: [PATCH 02/17] cob: check fid part of key in m0_cob_iterator_get() too

m0_cob_iterator_next() checks fid's of old and new positions
and if fid's differ, it returns -ENOENT. m0_cob_iterator_get()
has been modified to act in the same manner.
Call tracing has been added to both these functions.
---
 cob/cob.c | 33 ++++++++++++++++++++++++++++++---
 1 file changed, 30 insertions(+), 3 deletions(-)

diff --git a/cob/cob.c b/cob/cob.c
index 6cd77db..c4b8cc6 100644
--- a/cob/cob.c
+++ b/cob/cob.c
@@ -1002,18 +1002,45 @@ M0_INTERNAL int m0_cob_iterator_init(struct m0_cob *cob,
 
 M0_INTERNAL int m0_cob_iterator_get(struct m0_cob_iterator *it)
 {
-	return m0_db_cursor_get(&it->ci_cursor, &it->ci_pair);
+	int rc;
+	M0_ENTRY("[%lx:%lx]/%.*s", 
+		 (long)it->ci_key->cnk_pfid.f_container,
+		 (long)it->ci_key->cnk_pfid.f_key,
+		 m0_bitstring_len_get(&it->ci_key->cnk_name),
+		 (char *)m0_bitstring_buf_get(&it->ci_key->cnk_name));
+
+	rc = m0_db_cursor_get(&it->ci_cursor, &it->ci_pair);
+	if (rc == 0 && !m0_fid_eq(&it->ci_key->cnk_pfid, it->ci_cob->co_fid))
+		rc = -ENOENT;
+
+	M0_LEAVE("[%lx:%lx]/%.*s, rc: %d", 
+		 (long)it->ci_key->cnk_pfid.f_container,
+		 (long)it->ci_key->cnk_pfid.f_key,
+		 m0_bitstring_len_get(&it->ci_key->cnk_name),
+		 (char *)m0_bitstring_buf_get(&it->ci_key->cnk_name),
+		 rc);
+	return rc;
 }
 
 M0_INTERNAL int m0_cob_iterator_next(struct m0_cob_iterator *it)
 {
 	int rc;
+	M0_ENTRY("[%lx:%lx]/%.*s", 
+		 (long)it->ci_key->cnk_pfid.f_container,
+		 (long)it->ci_key->cnk_pfid.f_key,
+		 m0_bitstring_len_get(&it->ci_key->cnk_name),
+		 (char *)m0_bitstring_buf_get(&it->ci_key->cnk_name));
 
 	rc = m0_db_cursor_next(&it->ci_cursor, &it->ci_pair);
-
 	if (rc == 0 && !m0_fid_eq(&it->ci_key->cnk_pfid, it->ci_cob->co_fid))
-		return -ENOENT;
+		rc = -ENOENT;
 
+	M0_LEAVE("[%lx:%lx]/%.*s, rc: %d", 
+		 (long)it->ci_key->cnk_pfid.f_container,
+		 (long)it->ci_key->cnk_pfid.f_key,
+		 m0_bitstring_len_get(&it->ci_key->cnk_name),
+		 (char *)m0_bitstring_buf_get(&it->ci_key->cnk_name),
+		 rc);
 	return rc;
 }
 
-- 
1.8.3.2

